File: automedal/agent/tools/subagent.py
"""spawn_subagent tool — runs a focused sub-task in a fresh kernel.

Primary use: Researcher fanning out 3 arxiv queries in parallel via
asyncio.gather. The subagent inherits the parent's provider but gets a
fresh message history and a tool allowlist intersected against the
parent's available tools (subagent cannot escalate).
"""

from __future__ import annotations

from typing import Iterable

from automedal.agent.tools.base import Tool, ToolResult, make_tool


_DEFAULT_ALLOWLIST = ("read_file", "grep", "list_dir", "recall", "arxiv_search")


def make_subagent_tool(
    *,
    provider,
    parent_tools: list[Tool],
    events,
    depth: int = 0,
    max_depth: int = 2,
    default_max_steps: int = 20,
    default_allowlist: Iterable[str] = _DEFAULT_ALLOWLIST,
) -> Tool:
    """Build a `spawn_subagent` tool bound to the given provider + parent toolset.

    The returned tool, when called, builds a fresh AgentKernel with the
    subset of `parent_tools` named in `tools` (or the default allowlist),
    runs it on the supplied `prompt`, and returns the assistant's final text.
    """
    by_name = {t.name: t for t in parent_tools}
    default_set = tuple(default_allowlist)

    async def _spawn(
        prompt: str,
        tools: list[str] | None = None,
        max_steps: int = default_max_steps,
        label: str = "subagent",
    ) -> ToolResult:
        if depth >= max_depth:
            return ToolResult(
                f"error: subagent depth cap reached (depth={depth}, max={max_depth})",
                ok=False,
            )
        wanted = tuple(tools) if tools else default_set
        allowed = [by_name[n] for n in wanted if n in by_name]
        if not allowed:
            return ToolResult(
                f"error: none of the requested tools {wanted!r} are available "
                f"to the parent (parent has {sorted(by_name)})",
                ok=False,
            )

        # Local import to avoid a circular dep (kernel imports tools, tools
        # don't normally know about the kernel).
        from automedal.agent.kernel import AgentKernel

        sink = events.child_subagent(label=label) if events is not None else None
        if sink is not None:
            sink.subagent_start(label=label, prompt_preview=prompt[:160])

        sub_tools = list(allowed)
        # Allow nested subagents up to max_depth
        sub_tools.append(make_subagent_tool(
            provider=provider, parent_tools=allowed, events=sink,
            depth=depth + 1, max_depth=max_depth,
            default_max_steps=default_max_steps, default_allowlist=default_set,
        ))

        kernel = AgentKernel(
            provider=provider,
            system_prompt=(
                "You are a focused sub-agent. Complete the requested sub-task "
                "with the tools available, then summarize the result in your "
                "final assistant message. Do not chat — be terse and factual."
            ),
            tools=sub_tools,
            events=sink,
            max_steps=max_steps,
        )
        report = await kernel.run(prompt)

        if sink is not None:
            sink.subagent_end(label=label, ok=(report.stop == "assistant_done"),
                              result_preview=(report.final_text or "")[:160])
            # Do NOT call sink.close() — child sinks share file handles with the parent.

        if report.stop == "assistant_done":
            return ToolResult(report.final_text or "(empty)", ok=True)
        return ToolResult(
            f"subagent stopped with {report.stop}: {report.error or ''}",
            ok=False,
        )

    return make_tool(
        name="spawn_subagent",
        description=(
            "Spawn a focused sub-agent on `prompt`. The sub-agent gets a "
            "restricted tool allowlist (intersected against the parent's). "
            "Use for parallel fanout (e.g., 3 arxiv queries in parallel)."
        ),
        schema={
            "type": "object",
            "properties": {
                "prompt":    {"type": "string"},
                "tools":     {
                    "type": "array",
                    "items": {"type": "string"},
                    "description": "Tool name allowlist; subset of the parent's tools.",
                },
                "max_steps": {"type": "integer", "default": default_max_steps,
                              "minimum": 1, "maximum": 50},
                "label":     {"type": "string", "default": "subagent"},
            },
            "required": ["prompt"],
        },
        fn=_spawn,
    )

File: internal/agent/subagent.go
// spawn_subagent — runs a focused sub-task in a fresh AgentKernel.
//
// Lives in the agent package (not internal/agent/tools) to avoid the
// cycle: kernel.go → tools, so tools cannot import the kernel. Phases
// build this tool and add it to the parent kernel's toolset.
//
// Mirrors automedal/agent/tools/subagent.py.
package agent

import (
	"context"
	"fmt"
	"sort"

	"github.com/Flameingmoy/automedal/internal/agent/tools"
)

var defaultSubagentAllowlist = []string{"read_file", "grep", "list_dir", "recall", "arxiv_search"}

// SubagentConfig captures the knobs for MakeSubagentTool.
type SubagentConfig struct {
	Chat            ChatStreamFunc
	ParentTools     []tools.Tool
	Events          *EventSink
	Depth           int
	MaxDepth        int
	DefaultMaxSteps int
	Allowlist       []string // overrides the built-in default if non-nil
}

// MakeSubagentTool returns a `spawn_subagent` Tool bound to the parent's
// provider + toolset. The sub-agent gets a restricted allowlist and a
// fresh message history.
func MakeSubagentTool(cfg SubagentConfig) tools.Tool {
	if cfg.MaxDepth <= 0 {
		cfg.MaxDepth = 2
	}
	if cfg.DefaultMaxSteps <= 0 {
		cfg.DefaultMaxSteps = 20
	}
	defaultSet := cfg.Allowlist
	if defaultSet == nil {
		defaultSet = defaultSubagentAllowlist
	}

	byName := make(map[string]tools.Tool, len(cfg.ParentTools))
	for _, t := range cfg.ParentTools {
		byName[t.Name] = t
	}

	desc := "Spawn a focused sub-agent on `prompt`. The sub-agent gets a " +
		"restricted tool allowlist (intersected against the parent's). " +
		"Use for parallel fanout (e.g., 3 arxiv queries in parallel)."
	schema := map[string]any{
		"type": "object",
		"properties": map[string]any{
			"prompt": map[string]any{"type": "string"},
			"tools": map[string]any{
				"type":        "array",
				"items":       map[string]any{"type": "string"},
				"description": "Tool name allowlist; subset of the parent's tools.",
			},
			"max_steps": map[string]any{"type": "integer", "default": cfg.DefaultMaxSteps, "minimum": 1, "maximum": 50},
			"label":     map[string]any{"type": "string", "default": "subagent"},
		},
		"required": []string{"prompt"},
	}

	run := func(ctx context.Context, args map[string]any) (tools.ToolResult, error) {
		prompt := tools.StrArg(args, "prompt", "")
		if prompt == "" {
			return tools.MissingArg("prompt"), nil
		}
		if cfg.Depth >= cfg.MaxDepth {
			return tools.Error("error: subagent depth cap reached (depth=%d, max=%d)", cfg.Depth, cfg.MaxDepth), nil
		}
		label := tools.StrArg(args, "label", "subagent")
		maxSteps := tools.IntArg(args, "max_steps", cfg.DefaultMaxSteps)
		if maxSteps < 1 {
			maxSteps = 1
		}
		if maxSteps > 50 {
			maxSteps = 50
		}

		wanted := tools.StrListArg(args, "tools", nil)
		if len(wanted) == 0 {
			wanted = defaultSet
		}
		allowed := make([]tools.Tool, 0, len(wanted))
		for _, n := range wanted {
			if t, ok := byName[n]; ok {
				allowed = append(allowed, t)
			}
		}
		if len(allowed) == 0 {
			parentNames := make([]string, 0, len(byName))
			for n := range byName {
				parentNames = append(parentNames, n)
			}
			sort.Strings(parentNames)
			return tools.Error("error: none of the requested tools %v are available to the parent (parent has %v)",
				wanted, parentNames), nil
		}

		var sink *EventSink
		if cfg.Events != nil {
			sink = cfg.Events.ChildSubagent(label)
			sink.SubagentStart(label, prompt)
		}

		// Allow nested subagents up to max_depth.
		subTools := make([]tools.Tool, 0, len(allowed)+1)
		subTools = append(subTools, allowed...)
		subTools = append(subTools, MakeSubagentTool(SubagentConfig{
			Chat:            cfg.Chat,
			ParentTools:     allowed,
			Events:          sink,
			Depth:           cfg.Depth + 1,
			MaxDepth:        cfg.MaxDepth,
			DefaultMaxSteps: cfg.DefaultMaxSteps,
			Allowlist:       defaultSet,
		}))

		k := NewKernel(
			cfg.Chat,
			"You are a focused sub-agent. Complete the requested sub-task "+
				"with the tools available, then summarize the result in your "+
				"final assistant message. Do not chat — be terse and factual.",
			subTools,
			sink,
		)
		k.MaxSteps = maxSteps
		report := k.Run(ctx, prompt)

		ok := report.Stop == StopAssistantDone
		if sink != nil {
			sink.SubagentEnd(label, ok, report.FinalText)
			// Do NOT close — child sinks share file handles with the parent.
		}
		if ok {
			text := report.FinalText
			if text == "" {
				text = "(empty)"
			}
			return tools.Result(text), nil
		}
		return tools.Error("subagent stopped with %s: %s", report.Stop, report.Error), nil
	}

	return tools.Tool{Name: "spawn_subagent", Description: desc, Schema: schema, Run: run}
}

// silence unused import when tests strip log
var _ = fmt.Sprintf
